src/sonic_ax_impl/mibs/ietf/rfc3433.py

```diff
@@ -0,0 +1,402 @@
+"""
+RFC 3433 MIB implementation
+"""
+
+from enum import Enum, unique
+from bisect import bisect_right
+
+from swsssdk import SonicV2Connector
+from swsssdk import port_util
+from ax_interface import MIBMeta, MIBUpdater, ValueType, SubtreeMIBEntry
+from sonic_ax_impl import mibs
+
+@unique
+class EntitySensorDataType(int, Enum):
+    """
+    Enumeration of sensor data types according to RFC3433
+    (https://tools.ietf.org/html/rfc3433)
+    """
+
+    OTHER      = 1
+    UNKNOWN    = 2
+    VOLTS_AC   = 3
+    VOLTS_DC   = 4
+    AMPERES    = 5
+    WATTS      = 6
+    HERTZ      = 7
+    CELSIUS    = 8
+    PERCENT_RH = 9
+    RPM        = 10
+    CMM        = 11
+    TRUTHVALUE = 12
+
+
+@unique
+class EntitySensorDataScale(int, Enum):
+    """
+    Enumeration of sensor data scale types according to RFC3433
+    (https://tools.ietf.org/html/rfc3433)
+    """
+
+    YOCTO = 1
+    ZEPTO = 2
+    ATTO  = 3
+    FEMTO = 4
+    PICO  = 5
+    NANO  = 6
+    MICRO = 7
+    MILLI = 8
+    UNITS = 9
+    KILO  = 10
+    MEGA  = 11
+    GIGA  = 12
+    TERA  = 13
+    EXA   = 14
+    PETA  = 15
+    ZETTA = 16
+    YOTTA = 17
+
+
+@unique
+class EntitySensorStatus(int, Enum):
+    """
+    Enumeration of sensor operational status according to RFC3433
+    (https://tools.ietf.org/html/rfc3433)
+    """
+
+    OK             = 1
+    UNAVAILABLE    = 2
+    NONOPERATIONAL = 3
+
+
+@unique
+class EntitySensorValueRange(int, Enum):
+    """
+    Range of EntitySensorValue field defined by RFC 3433
+    """
+
+    MIN = -1E9
+    MAX = 1E9
+
+
+class SensorInterface:
+    """
+    Sensor interface.
+    Sensor should define SCALE, TYPE, PRECISION
+    """
+
+    SCALE = None
+    TYPE = None
+    PRECISION = None
+
+
+    @classmethod
+    def mib_values(cls, raw_value, converter=lambda x: x):
+        """
+        :param: cls: class instance
+        :param: value: sensor's value as is from DB
+        :param: converter: optional converter for a value in case it
+        is needed to convert value from one unit to anther
+        :return: value converted for MIB
+        """
+
+        type_ = cls.TYPE
+        scale = cls.SCALE
+        precision = cls.PRECISION
+
+        try:
+            value = converter(float(raw_value))
```

Comment:
>converter [](start = 20, length = 9)

if converter raises a ValueError, should we treat it as a bug?

Reply:
If float() raises ValueError than we cannot parse the value(It could be "N/A" in DB), then we teat sensor as UNAVAILABLE.
If converter raises a ValueError - we cannot convert it, (e.g if sensor value in DB is 0 and MIB should return this value in logarithmic scale, then converter will raise a ValueError and we will treat the sensor as UNAVAILABLE)
E.g:
>>> math.log(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: math domain error

Reply:
I think the 'logarithmic' convert is a perfect example for my argument. If the raw value is 0, we should treat as sensor status available, but converter is a buggy one. We don't want to treat it as same as sensor status missing. Log it as a bug or propagate the error to downstream will help expose the bug.

---
In reply to: [200947198](https://github.com/Azure/sonic-snmpagent/pull/69#discussion_r200947198) [](ancestors = 200947198)

Reply:
In case converter() call raises exception (ValueError or ArithmeticError) it is propaget to update_data() and an error is logged, entry skipped (since we don't know what to put into MIB, 0?)
I believe a good converter implementation for logarithmic convert should return float('-inf') in case original value is 0. In this case the code will correctly compare float('-inf') < EntitySensorValueRange.MIN, and set the value for mib to EntitySensorValueRange.MIN, so user will know that this sensor overflows and that this sensor is in logarithmic scale (according to TYPE field in MIB). Does it make sense?

Reply:
Make sense. Thanks for fixing and explanation.

---
In reply to: [201421302](https://github.com/Azure/sonic-snmpagent/pull/69#discussion_r201421302) [](ancestors = 201421302)